### PR TITLE
[DBZ-PGYB][yugabyte/yugabyte-db#30733] Fix null Kafka key when connector resumes after table drop

### PR DESCRIPTION
**Pull request by shishir2001-yb**
## Problem

`PgOutputMessageDecoder.handleRelationMessage()` resolves primary key columns by querying the live database via `connection.readPrimaryKeyNames()`. When the connector is stopped, rows are inserted (accumulating in WAL), the table is dropped, and the walsender is killed, upon restart the connector establishes a fresh replication connection, receives a new relation message for the dropped table, and the PK query returns empty. This results in all pending Kafka records being produced with a **null key**.

The WAL relation message already carries a per-column `flags` byte that identifies PK columns, but the existing code reads it without using it for PK resolution.

## Solution

Refactor PK resolution in `handleRelationMessage()` to use the relation message `flags` byte instead of querying the database.

### Implementation Details

1. **Use Relation message flags for DEFAULT/INDEX/CHANGE replica identities**
   - Derive `primaryKeyColumns` from the relation message `flags` byte (`flags & 1 == 1` means PK column)
   - No out-of-band DB query needed — works even if the table has been dropped

2. **Fall back to DB query for FULL/NOTHING replica identities**
   - `FULL` sets all flags to 1, `NOTHING` sets all to 0, flags are ambiguous for PK resolution
   - Query `connection.readPrimaryKeyNames()` only for these cases

3. **Backward-compatible CHANGE fallback**
   - YugabyteDB's `yboutput` may not set flags for CHANGE identity in older versions (YB#22555)
   - If CHANGE yields an empty PK list from flags, fall back to a DB query

4. **Remove unnecessary out-of-band DB queries**
   - Removed queries for `columnDefaults` and `columnOptionality`, use safe defaults (`optional=true`, `hasDefault=false`) instead
   - Extracted `queryPrimaryKeysFromDatabase()` and `parseReplicaIdentity()` helper methods

5. **Integration tests (`YBHandleRelationMessageIT`)**
   - PK resolution tests for DEFAULT, FULL, and CHANGE replica identities
   - Table-drop + connector restart scenario: stop connector, insert rows, drop table, kill walsender, restart, verify all records have non-null keys
   - Red-green verified: test fails with null key on original code, passes with the fix
 

### Trade-off: Column metadata defaults
The previous code queried the database for per-column optionality (`IS NOT NULL` constraints) and default value expressions. With this change, all columns default to `optional=true` and `hasDefault=false`. This means the connector's schema metadata for columns will not reflect `NOT NULL` constraints or default values from the database. This is an acceptable trade-off because:
- These metadata fields do not affect the Kafka record key (the core fix)
- The relation message protocol does not carry optionality or default info
- Querying the DB for this metadata is what causes the failure when the table is dropped

## Test plan

```
YBHandleRelationMessageIT#shouldResolvePkWithReplicaIdentityDefault
YBHandleRelationMessageIT#shouldResolvePkWithReplicaIdentityFull
YBHandleRelationMessageIT#shouldResolvePkWithReplicaIdentityChange
YBHandleRelationMessageIT#shouldResolvePkAfterTableDropWithReplicaIdentityDefault
YBHandleRelationMessageIT#shouldResolvePkAfterTableDropWithReplicaIdentityFull
YBHandleRelationMessageIT#shouldResolvePkAfterTableDropWithReplicaIdentityChange

```

---

> [!NOTE]
> **Medium Risk**
> Changes primary-key resolution during pgoutput relation decoding and removes DB-derived column metadata, which can affect emitted record keys and schema details during streaming/restarts. Adds new Yugabyte-focused integration tests covering multiple replica identities and restart-after-drop scenarios.
> 
> **Overview**
> Fixes a restart edge case where pending WAL events could be emitted with **null Kafka keys** after a table is dropped by changing `PgOutputMessageDecoder.handleRelationMessage()` to derive primary-key columns from the replication relation message `flags` (with DB-metadata fallback for `FULL/NOTHING`, and a `CHANGE` fallback for older `yboutput`).
> 
> Reduces reliance on out-of-band metadata by dropping per-column default/optionality lookups and defaulting columns to `optional=true` / `hasDefault=false`, and adds `YBHandleRelationMessageIT` to verify non-null keys across replica identities and a stop/insert/drop/restart scenario.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4052df0b50496c94087586f6e863da07df6a861b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YBHandleRelationMessageIT.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright Debezium Authors.
```
Comment (shishir2001-yb) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YBHandleRelationMessageIT.java]: Are we supposed to add Yugabyte here? :)

Comment (suranjan) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/YBHandleRelationMessageIT.java]: yes

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java [hunk 2]
```diff
@@ -400,21 +408,45 @@ private void handleRelationMessage(ByteBuffer buffer, TypeRegistry typeRegistry)
...
 
+    /**
+     * Queries the database for primary key columns of the given table.
+     * Falls back to unique indices if no primary keys are found.
+     */
+    private List<String> queryPrimaryKeysFromDatabase(TableId tableId) throws SQLException {
+        final DatabaseMetaData databaseMetadata = connection.connection().getMetaData();
+        List<String> primaryKeyColumns = connection.readPrimaryKeyNames(databaseMetadata, tableId);
+        if (primaryKeyColumns == null || primaryKeyColumns.isEmpty()) {
+            LOGGER.warn("Primary keys are not defined for table '{}', defaulting to unique indices", tableId);
```
Comment (suranjan) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:430]: Will it always log warning for non primary key tables?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java [hunk 3]
```diff
@@ -400,21 +408,45 @@ private void handleRelationMessage(ByteBuffer buffer, TypeRegistry typeRegistry)
         // to reflect the actual primary key state at time `t0`.
         primaryKeyColumns.retainAll(columnNames);
 
+        LOGGER.trace("Final primaryKeyColumns for '{}.{}': {}", schemaName, tableName, primaryKeyColumns);
+
         Table table = resolveRelationFromMetadata(new PgOutputRelationMetaData(relationId, schemaName, tableName, columns, primaryKeyColumns));
         if (YugabyteDBServer.isEnabled()) {
             decoderContext.getSchema().applySchemaChangesForTableWithReplicaIdentity(relationId, table, replicaIdentityId);
-        } else {
+        }
```
Comment (suranjan) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:416]: unncessary changes

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 2 above)
Comment (shishir2001-yb) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:430]: Not always. This path is only reached when replica identity is FULL or NOTHING, or as a CHANGE identity fallback

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 3 above)
Comment (shishir2001-yb) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:416]: it was done by Auto formatter. Let it be there so it won't pop up again

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 2 above)
Comment (suranjan) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:430]: Isn't that a log spew?

**Review by suranjan — APPROVED** [review]

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 2 above)
Comment (shishir2001-yb) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:430]: As discussed offline this is existing behaviour so should be fine
